File: library/src/main/java/cn/wandersnail/common/http/upload/UploadWorker.java
package cn.wandersnail.common.http.upload;

import java.io.File;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.List;
import java.util.Map;

import cn.wandersnail.common.http.util.HttpUtils;
import cn.wandersnail.common.http.util.SchedulerUtils;
import io.reactivex.disposables.Disposable;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;

/**
 * 上传执行
 * <p>
 * date: 2019/8/23 21:41
 * author: zengfansheng
 */
public class UploadWorker<T> implements Disposable {
    private final UploadObserver<T> observer;

    public UploadWorker(UploadInfo<T> info, UploadListener<T> listener) {
        observer = new UploadObserver<>(info, listener);
        Retrofit.Builder builder = new Retrofit.Builder();
        if (info.client != null) {
            builder.client(HttpUtils.initHttpsClient(true, new OkHttpClient.Builder()).build());
        }
        UploadService service = builder.addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .baseUrl(info.getBaseUrl())
                .build()
                .create(UploadService.class);
        MultipartBody.Builder bodyBuilder = new MultipartBody.Builder();
        if (info.paramParts != null) {
            for (Map.Entry<String, String> entry : info.paramParts.entrySet()) {
                bodyBuilder.addFormDataPart(entry.getKey(), entry.getValue());
            }
        }
        for (Map.Entry<String, List<File>> entry : info.fileParts.entrySet()) {
            try {
                List<File> files = entry.getValue();
                for (File file : files) {
                    MultipartBody.Part part = MultipartBody.Part.createFormData(entry.getKey(),
                            URLEncoder.encode(file.getName(), "utf-8"),
                            new ProgressRequestBody(MediaType.parse("multipart/form-data"), entry.getKey(),
                                    file, observer));
                    bodyBuilder.addPart(part);
                }
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
            }
        }
        service.upload(info.url, bodyBuilder.build())
                .compose(SchedulerUtils.applyGeneralObservableSchedulers())
                .subscribe(observer);
    }

    @Override
    public void dispose() {
        observer.dispose();
    }

    @Override
    public boolean isDisposed() {
        return observer.isDisposed();
    }
    
    public void cancel() {
        dispose();
    }
}

File: library/src/main/java/com/snail/java/network/upload/UploadWorker.kt
package com.snail.java.network.upload

import io.reactivex.disposables.Disposable
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import java.net.URLEncoder

/**
 * 上传执行者
 *
 * date: 2019/2/28 18:21
 * author: zengfansheng
 */
class UploadWorker<T> @JvmOverloads internal constructor(info: UploadInfo<T>, listener: UploadListener<T>? = null) : Disposable {
    private val observer = UploadObserver(info, listener)

    init {
        val builder = Retrofit.Builder()
        if (info.client != null) {
            builder.client(info.client)
        }
        val service = builder.addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .baseUrl(info.baseUrl)
            .build()
            .create(UploadService::class.java)
        val parts = HashMap<String, @JvmSuppressWildcards RequestBody>()
        info.paramParts.entries.forEach { parts[it.key] = RequestBody.create(null, it.value) }
        info.fileParts.entries.forEach {
            MultipartBody.Part.createFormData(it.key, URLEncoder.encode(it.value.name, "utf-8"),
                ProgressRequestBody(MediaType.parse("multipart/form-data"), it.key, it.value, observer))
        }
        service.upload(info.url, parts).subscribe(observer)
    }

    override fun dispose() {
        observer.dispose()
    }

    override fun isDisposed(): Boolean {
        return observer.isDisposed
    }

    fun cancel() {
        dispose()
    }
}